rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -250,129 +234,95 @@ object SpatialMatcher extends SpatialMatcherInstances {
 
     result
   }
 
-  def listMatch[T](tlist: Seq[T],
-                   plist: Seq[T],
-                   merger: (Par, T) => Par,
-                   varLevel: Option[Int],
-                   wildcard: Boolean)(implicit lf: HasLocallyFree[T],
-                                      sm: SpatialMatcher[T, T]): NonDetFreeMapWithCost[Unit] =
-    (tlist, plist) match {
-      // Handle the remainder.
-      case (rem, Nil) =>
-        varLevel match {
-          case None =>
-            // If there is a wildcard, we succeed.
-            // We also succeed if there isn't but the remainder is empty.
-            if (wildcard || rem == Nil)
-              NonDetFreeMapWithCost.pure(Unit)
-            else
-              // This should be prevented by the length checks.
-              NonDetFreeMapWithCost.emptyMap[Unit]
-          // If there's a capture variable, we prefer to add things to that rather than throw them
-          // away.
-          case Some(level) => {
-            handleRemainder(rem, level, merger, wildcard)
-          }
-        }
-      // Try to find a match for a single pattern.
-      case (targets, pattern +: prem) => {
-        if (!lf.connectiveUsed(pattern)) {
-          possiblyRemove(pattern, targets) match {
-            case None => NonDetFreeMapWithCost.liftF(Stream.empty)
-            case Some(filtered) =>
-              listMatch(filtered, prem, merger, varLevel, wildcard)
-          }
-        } else {
-          for {
-            trem        <- listMatchItem(targets, pattern, spatialMatch[T, T])
-            forcedYield <- listMatch(trem, prem, merger, varLevel, wildcard)
-          } yield forcedYield
-        }
-      }
-    }
-
-  private def handleRemainder[T](rem: Seq[T],
-                                 level: Int,
-                                 merger: (Par, T) => Par,
+  private[this] def listMatch[T](targets: Seq[T],
+                                 patterns: Seq[T],
+                                 merger: (Par, Seq[T]) => Par,
+                                 varLevel: Option[Int],
                                  wildcard: Boolean)(
       implicit lf: HasLocallyFree[T],
-      sm: SpatialMatcher[T, T]): NonDetFreeMapWithCost[Unit] = {
-    // This function is essentially an early terminating left fold.
-    @tailrec
-    def foldRemainder(remainder: Seq[T], p: Par): NonDetFreeMapWithCost[Par] =
-      remainder match {
-        case Nil => NonDetFreeMapWithCost.pure(p)
-        case item +: rem =>
-          if (lf.locallyFree(item, 0).isEmpty)
-            foldRemainder(rem, merger(p, item))
-          else if (wildcard)
-            foldRemainder(rem, p)
-          else
-            NonDetFreeMapWithCost.emptyMap[Par]
+      sm: SpatialMatcher[T, T]): OptionalFreeMapWithCost[Unit] = {
+
+    sealed trait Pattern
+    case class Term(term: T)         extends Pattern
+    case class Remainder(level: Int) extends Pattern
+
+    val remainderPatterns: Seq[Pattern] = varLevel.fold(
```

Comment:
Why are you creating as many remainder patterns as the `varLevel`?

Reply:
I know.

Reply:
He's using fold to pattern match on the option.